File: P5/QytetetRuby/lib/vista_textual_qytetet.rb
#encoding:utf-8
#qyt = Qytetet.instance
require_relative 'controladorqytetet'
require_relative 'opcion_menu'
require_relative 'qytetet'
require_relative 'jugador'

module VistaTextualQytetet
    class VistaTextualQytetet
      def initialize()
        @modelo = ModeloQytetet::Qytetet.instance
        @controlador = ControladorQytetet::ControladorQytetet.instance
      end
      public
      def obtenerNombreJugadores()
        jugadores = Array.new 
        puts "Indica el numero de jugadores(de 2 a 4)"
        #meter el numero de jugadores
        i = gets.chomp
        i.to_i.times do |j|
          puts "Introduce un nombre: \n"
          s = gets.chomp
          jugadores << s
        end
        #bucle en el que se añaden los jugadores
        return jugadores
      end
      
      def elegirCasilla(opcionMenu)
        casillasValidas = Array.new
        casillasValidas = @controlador.obtenerCasillasValidas(opcionMenu)
        if(casillasValidas.empty?)
          puts "No hay casillas validas"
          return -1
        else
          convertidos = Array.new
          puts "Elige el numero de casilla: "
          for i in casillasValidas
            puts i.to_s + " "
            convertidos << i.to_s
          end
          return leerValorCorrecto(convertidos).to_i
        end
      end
      
      def leerValorCorrecto(valoresCorrectos)
        #leer en un string lo que el usuario escribe por consola
        s = gets.chomp
        #comprobar si el valor pertenece a la lista de valores
        while(!valoresCorrectos.include?(s))
          puts "Error, el elemento introducido no pertenece a la lista de opciones.\nIntroduzca un valor correcto: "
          s = gets.chomp
        end
        return s
      end
      
      def elegirOperacion()
        listaOperaciones = Array.new
        listaOperaciones = @controlador.obtenerOperacionesDeJuegoValidas()
        convertidos = Array.new
        
        if(!@modelo.jugadores.empty?)
          puts "Es el turno de #{@modelo.jugadorActual.nombre}"
        end
        
        for i in listaOperaciones
          puts i.to_s + ": " + ControladorQytetet::OpcionMenu[i].to_s
        end
        
        for i in listaOperaciones
          convertidos << i.to_s
        end
        
        aux = leerValorCorrecto(convertidos)
        return aux.to_i
      end
      
      def main()
        ui = VistaTextualQytetet.new
        @controlador.nombreJugadores = ui.obtenerNombreJugadores()
        operacionElegida=0
        casillaElegida=0
        necesitaElegirCasilla=false
        while(true)
          operacionElegida = ui.elegirOperacion()
          necesitaElegirCasilla = @controlador.necesitaElegirCasilla(operacionElegida)
          if(necesitaElegirCasilla)
            casillaElegida=ui.elegirCasilla(operacionElegida)
          end
          if(!necesitaElegirCasilla || casillaElegida >=0)
            puts "#{@controlador.realizarOperacion(operacionElegida, casillaElegida)}"
          end
        end 
      end     
    end
    
    
    
    
    a = VistaTextualQytetet.new
    a.main
end

File: P5/QytetetJava/src/vistatextualqytetet/VistaTextualQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package vistatextualqytetet;
import controladorqytetet.ControladorQytetet;
import controladorqytetet.OpcionMenu;
import modeloqytetet.Qytetet;
import java.util.ArrayList;
import modeloqytetet.Jugador;
import java.util.Scanner;

/**
 *
 * @author pedrohueneja
 */
public class VistaTextualQytetet {
    private static Qytetet modelo = Qytetet.getInstance();
    private static ControladorQytetet controlador = ControladorQytetet.getInstance();
    Scanner in = new Scanner(System.in);
    
    
    public ArrayList<String> obtenerNombreJugadores(){
        ArrayList<String> _jugadores = new ArrayList<>();
        System.out.println("Indica el número de jugadores (de 2 a " + modelo.MAX_JUGADORES + "): ");
        String s = in.nextLine();
        
        for(int i=0; i < Integer.parseInt(s); i++){
            System.out.println("Introduce un nombre: ");
            String nombre = in.nextLine();
            _jugadores.add(nombre);
        }
        return _jugadores;
        
    }
    public int elegirCasilla(int opcionMenu){
        ArrayList<Integer> casillasValidas = controlador.obtenerCasillasValidas(opcionMenu);
        if(casillasValidas.isEmpty()){
            return -1;
        }
        else{
            //Mostrar Lista
            //Pasar a string la lista
            ArrayList<String> convertidos = new ArrayList<String>();
             System.out.println("Elige el numero de casilla: ");
            for(Integer i: casillasValidas){
                System.out.println(i + "  ");
                convertidos.add(Integer.toString(i));
            }
            return Integer.parseInt(leerValorCorrecto(convertidos));            
        }
    }
    
    public String leerValorCorrecto(ArrayList<String> valoresCorrectos){
        //leer en un string lo que el usuario escribe por consola
        String s = in.nextLine();
        //comprobar si el valor pertenece a la lista de valores
        while(!valoresCorrectos.contains(s)){
            System.out.print("Error, el elemento introducido no pertenece a la lista de opciones.\nIntroduzca un valor correcto: ");
            s = in.nextLine();
        }
        return s;
    }
    
    public int elegirOperacion(){
        ArrayList<Integer> listaOpciones = controlador.obtenerOperacionesJuegoValidas();
        ArrayList<String> convertidos = new ArrayList<String>();
        if(!modelo.getJugadores().isEmpty())
        System.out.println("Es el turno de " + modelo.getJugadorActual().getNombre());
        //Mostrar las opciones disponibles:
        for(int i: listaOpciones){
            System.out.println(i + ": " + OpcionMenu.values()[i]);
        }
        
        for(Integer i: listaOpciones){
            convertidos.add(Integer.toString(i));
        }
        String aux = leerValorCorrecto(convertidos);
        return Integer.parseInt(aux);
    }
    
     public static void main(String[] args) {
        VistaTextualQytetet ui = new VistaTextualQytetet();        
        controlador.setNombreJugadores(ui.obtenerNombreJugadores());
        int operacionElegida,casillaElegida = 0;
        boolean necesitaElegirCasilla;
        
        do{
            operacionElegida = ui.elegirOperacion();
            necesitaElegirCasilla = controlador.necesitaElegirCasilla(operacionElegida);
            if(necesitaElegirCasilla)
                casillaElegida = ui.elegirCasilla(operacionElegida);
            if(!necesitaElegirCasilla || casillaElegida >= 0)
                System.out.println(controlador.realizarOperacion(operacionElegida, casillaElegida));
        }while(true);
    }
}
